pkg/applicationserver/io/pubsub/redis/registry.go

```diff
@@ -0,0 +1,263 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package redis
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-redis/redis"
+	"github.com/gogo/protobuf/proto"
+	"go.thethings.network/lorawan-stack/pkg/applicationserver/io/pubsub"
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	ttnredis "go.thethings.network/lorawan-stack/pkg/redis"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/unique"
+)
+
+var (
+	errInvalidFieldmask   = errors.DefineInvalidArgument("invalid_fieldmask", "invalid fieldmask")
+	errInvalidIdentifiers = errors.DefineInvalidArgument("invalid_identifiers", "invalid identifiers")
+)
+
+// appendImplicitPubSubGetPaths appends implicit ttnpb.ApplicationPubSub get paths to paths.
+func appendImplicitPubSubGetPaths(paths ...string) []string {
+	return append(append(make([]string, 0, 4+len(paths)),
+		"created_at",
+		"ids",
+		"updated_at",
+		"provider",
+	), paths...)
+}
+
+func applyPubSubFieldMask(dst, src *ttnpb.ApplicationPubSub, paths ...string) (*ttnpb.ApplicationPubSub, error) {
+	if dst == nil {
+		dst = &ttnpb.ApplicationPubSub{}
+	}
+	return dst, dst.SetFields(src, paths...)
+}
+
+// PubSubRegistry is a Redis PubSub registry.
+type PubSubRegistry struct {
+	Redis *ttnredis.Client
+}
+
+func (r *PubSubRegistry) allKey(ctx context.Context) string {
+	return r.Redis.Key("all")
+}
+
+func (r *PubSubRegistry) appKey(uid string) string {
+	return r.Redis.Key("uid", uid)
+}
+
+func (r *PubSubRegistry) idKey(appUID, id string) string {
```

Comment:
```suggestion
func (r *PubSubRegistry) uidKey(appUID, id string) string {
```

nitpick; this is really UID in context of PubSubs. This key uniquely identifies a PubSub.

Reply:
Also changed `makeUIDKeyFunc`.